port/unix/omrsysinfo.c

```diff
@@ -3071,3 +3158,356 @@ omrsysinfo_os_kernel_info(struct OMRPortLibrary *portLibrary, struct OMROSKernel
+	}
+	rc = 0;
+
+_end:
+	if (NULL != cgroupFile) {
+		fclose(cgroupFile);
+	}
+	if (0 != rc) {
+		freeCgroupEntries(portLibrary, cgEntryList);
+		cgEntryList = NULL;
+	} else {
+		*cgroupEntryList = cgEntryList;
+	}
+	
+	return rc;
+}
+
+/**
+ * Read a file under a subsystem in cgroup hierarchy based on the format specified
+ *
+ * @param[in] portLibrary pointer to OMRPortLibrary
+ * @param[in] subsystem cgroup subsystem
+ * @param[in] fileName name of the file under cgroup subsystem
+ * @param[in] numItemsToRead number of items to be read from the file
+ * @param[in] format format to be used for reading the file
+ *
+ * @return 0 on successfully reading 'numItemsToRead' items from the file, negative error code on any error
+ */
+static int32_t
+readCgroupSubsystemFile(struct OMRPortLibrary *portLibrary, CgroupSubsystems subsystem, const char *fileName, int32_t numItemsToRead, const char *format, ...)
+{
+	char fileToRead[PATH_MAX];
+	char *cgroup = NULL;
+	FILE *file = NULL;
+	int32_t rc = 0;
+	va_list args;
+
+	if (!portLibrary->sysinfo_is_cgroup_limits_enabled(portLibrary)) {
+		rc = portLibrary->error_set_last_error_with_message(portLibrary, OMRPORT_ERROR_SYSINFO_CGROUP_LIMITS_DISABLED, "cgroup limits is disabled");
+		goto _end;
+	}
+	
+	cgroup = getCgroupNameForSubsystem(portLibrary, PPG_cgroupEntryList, subsystemNames[subsystem]);
+	if (NULL == cgroup) {
+		char errBuf[512];
+
+		portLibrary->str_printf(portLibrary, errBuf, sizeof(errBuf), "cgroup name for subsystem %s is not available", subsystemNames[MEMORY]);
+		rc = portLibrary->error_set_last_error_with_message(portLibrary, OMRPORT_ERROR_SYSINFO_CGROUP_NAME_NOT_AVAILABLE, errBuf);
+		goto _end;
+	}
+
+	portLibrary->str_printf(portLibrary, fileToRead, sizeof(fileToRead), "/sys/fs/cgroup/%s/%s/%s", subsystemNames[subsystem], cgroup, fileName);
+	file = fopen(fileToRead, "r");
+	if (NULL == file) {
+		portLibrary->error_set_last_error(portLibrary, errno, OMRPORT_ERROR_SYSINFO_CGROUP_MEMLIMIT_FILE_FOPEN_FAILED);
+		goto _end;
+	}
+
+	va_start(args, format);	
+	rc = vfscanf(file, format, args);
+	va_end(args);
+
+	if (numItemsToRead != rc) {
+		char errBuf[512];
+
+		portLibrary->str_printf(portLibrary, errBuf, sizeof(errBuf), "unexpcted format of file %s", fileName);
+		rc = portLibrary->error_set_last_error_with_message(portLibrary, OMRPORT_ERROR_SYSINFO_PROCESS_CGROUP_FILE_READ_FAILED, errBuf);
+		goto _end;
+	} else {
+		rc = 0;
+	}
+
+_end:
+	if (NULL != file) {
+		fclose(file);
+	}
+	return rc;
+}
+
+#endif /* defined(LINUX) */
+
+int32_t
+omrsysinfo_is_cgroup_limits_supported(struct OMRPortLibrary *portLibrary)
+{
+	int32_t rc = OMRPORT_ERROR_SYSINFO_CGROUP_UNSUPPORTED_PLATFORM;
+#if defined(LINUX)
+	char fileToTest[PATH_MAX];
+	struct statfs buf = {0};
+
+	if (NULL == PPG_cgroupEntryList) {
+		/* Check if tmpfs is mounted on /sys/fs/cgroup */
+		portLibrary->str_printf(portLibrary, fileToTest, sizeof(fileToTest), "/sys/fs/cgroup/");
+		rc = statfs(fileToTest, &buf);
```

Comment:
Why do the str_printf() here?  Couldn't the `"/sys/fs/cgroup/"` string be passed directly to `statsfs` saving the printf call and allowing the `fileToTest` buffer to be removed?